ion-schema/src/isl/ranges.rs

```diff
@@ -0,0 +1,471 @@
+use crate::isl::util::TimestampPrecision;
+use crate::{invalid_schema_error, invalid_schema_error_raw, isl_require};
+use crate::{IonSchemaResult, WriteToIsl};
+use ion_rs::element::writer::ElementWriter;
+use ion_rs::element::Element;
+use ion_rs::Decimal;
+use ion_rs::{IonType, IonWriter, Timestamp};
+use std::fmt::{Display, Formatter};
+
+/// The end of a [`Range`].
+#[derive(Debug, PartialEq, Clone)]
+pub enum Limit<T> {
+    Unbounded,
+    Closed(T),
+    Open(T),
+}
+
+impl<T: RangeOverType> Limit<T> {
+    /// Checks if a value is above this [`Limit`], assuming that this [`Limit`] is being used as the lower end of a [`Range`].
+    fn is_above<V: Into<T> + Clone>(&self, other: &V) -> bool {
+        let other = &other.clone().into();
+        match self {
+            Limit::Unbounded => true,
+            Limit::Open(this) => this > other,
+            Limit::Closed(this) => this >= other,
+        }
+    }
+
+    /// Checks if a value is below this [`Limit`], assuming that this [`Limit`] is being used as the upper end of a [`Range`].
+    fn is_below<V: Into<T> + Clone>(&self, other: &V) -> bool {
+        let other = &other.clone().into();
+        match self {
+            Limit::Unbounded => true,
+            Limit::Open(this) => this < other,
+            Limit::Closed(this) => this <= other,
+        }
+    }
+}
+impl<T: Display> Limit<T> {
+    fn fmt_for_display(&self, f: &mut Formatter<'_>, unbounded_text: &str) -> std::fmt::Result {
+        match self {
+            Limit::Unbounded => f.write_str(unbounded_text),
+            Limit::Closed(value) => value.fmt(f),
+            Limit::Open(value) => write!(f, "exclusive::{value}"),
+        }
+    }
+}
+
+/// Defines a type that can be "ranged over" using a [`Range`].
+pub trait RangeOverType: PartialEq + PartialOrd + Clone
+where
+    Self: Sized,
+{
+    /// When constructing a range, should an Open limit be adjusted into a closed limit?
+    /// Default implementation is to do nothing to the limit.
+    fn get_adjustment_for_open_lower_bound(self) -> Limit<Self> {
+        Limit::Open(self)
+    }
+    /// When constructing a range, should an Open limit be adjusted into a closed limit?
+    /// Default implementation is to do nothing to the limit.
+    fn get_adjustment_for_open_upper_bound(self) -> Limit<Self> {
+        Limit::Open(self)
+    }
+}
+
+impl RangeOverType for Decimal {}
+impl RangeOverType for Timestamp {}
+impl RangeOverType for TimestampPrecision {}
+impl RangeOverType for usize {
+    fn get_adjustment_for_open_lower_bound(self) -> Limit<Self> {
+        Limit::Closed(self + 1)
+    }
+    fn get_adjustment_for_open_upper_bound(self) -> Limit<Self> {
+        Limit::Closed(self - 1)
+    }
+}
```

Comment:
🗺️ The adjustment from `Open(n)` to `Closed(n ± 1)` is necessary for integer types in order to be able to make sure that the range is non-empty.

Reply:
Yes. I'll fix that.